pkg/handlers/v1/ec2.go

```diff
@@ -67,42 +70,115 @@ type ec2ChangedPropsARN struct {
+	PrivateIPs []string
+	PublicIPs  []string
+	Hostnames  []string
+}
+
 type ec2Transformer struct{}
 
-func (t ec2Transformer) Create(event awsConfigEvent) (Output, error) {
-	output, err := getBaseOutput(event.ConfigurationItem)
+func (t ec2Transformer) Create(event awsConfigEvent) ([]Output, error) {
+	outputs := []Output{}
+	baseOutput, err := getBaseOutput(event.ConfigurationItem)
 	if err != nil {
-		return Output{}, err
+		return []Output{}, err
 	}
-
 	// if a resource is created for the first time, there is no diff.
 	// just read the configuration
 	var config ec2Configuration
 	if err := json.Unmarshal(event.ConfigurationItem.Configuration, &config); err != nil {
-		return Output{}, err
+		return []Output{}, err
 	}
-	change := extractEC2NetworkInfo(&config)
-	change.ChangeType = added
-	output.Changes = append(output.Changes, change)
-	return output, nil
+
+	hasEni := false
+	for i := range config.NetworkInterfaces {
+		hasEni = true
+		output := baseOutput
+		change := Change{}
+		output.ChangeTime = config.NetworkInterfaces[i].Attachment.AttachTime
+		private, public, dns := extractNetworkInterfaceInfo(&config.NetworkInterfaces[i])
+		change.PrivateIPAddresses = append(change.PrivateIPAddresses, private...)
+		change.PublicIPAddresses = append(change.PublicIPAddresses, public...)
+		change.Hostnames = append(change.Hostnames, dns...)
+		change.ChangeType = added
+		output.Changes = append(output.Changes, change)
+		outputs = append(outputs, output)
+	}
+	if !hasEni {
+		return []Output{}, nil // Reject EC2s with no ENI
+	}
+
+	return outputs, nil
 }
 
-func (t ec2Transformer) Update(event awsConfigEvent) (Output, error) {
-	output, err := getBaseOutput(event.ConfigurationItem)
+func initializeAttachMapEntry(attachMap map[string]*eniIPList, attachTime string) {
+	attachMap[attachTime] = &eniIPList{}
+	attachMap[attachTime].PrivateIPs = make([]string, 0)
+	attachMap[attachTime].PublicIPs = make([]string, 0)
+	attachMap[attachTime].Hostnames = make([]string, 0)
+}
+
+func separateAddedChanges(baseOutput Output, addedChange Change, privateIPMap map[string]string, publicIPMap map[string]string, hostnameMap map[string]string) []Output {
+	outputs := make([]Output, 0)
+	attachMap := make(map[string]*eniIPList)
+	// separate IPs & hostnames by attachTime
+	for _, ip := range addedChange.PrivateIPAddresses {
+		attachTime := privateIPMap[ip]
+		if _, ok := attachMap[attachTime]; !ok {
+			initializeAttachMapEntry(attachMap, attachTime)
+		}
+		attachMap[attachTime].PrivateIPs = append(attachMap[attachTime].PrivateIPs, ip)
+	}
+	for _, ip := range addedChange.PublicIPAddresses {
+		attachTime := publicIPMap[ip]
+		if _, ok := attachMap[attachTime]; !ok {
+			initializeAttachMapEntry(attachMap, attachTime)
+		}
+		attachMap[attachTime].PublicIPs = append(attachMap[attachTime].PublicIPs, ip)
+	}
+	for _, hostname := range addedChange.Hostnames {
+		// the attachTime will always already be in the map, because adding the private IP(s)
```

Comment:
Nit on this comment: wouldn't it be because the public IPs have created an entry?

Reply:
The map key is the attachTime, and the private IP is the first to be added to a map entry, so it will create an entry with that attachTime. Based on the sample data I've seen, instances always have a private IP - although I'm not sure if that's always true, since I've only seen a small number of samples. The function `extractNetworkInterfaceInfo` seems to assume that there is at least 1 private IP, otherwise it returns no IPs or hostnames altogether.

Reply:
👍 I like that. Even after this amount of time I get thrown by corner cases on config events, so extra safety is always appreciated.